be/src/formats/parquet/file_writer.h

```diff
@@ -67,30 +79,108 @@ class ParquetBuildHelper {
 
     static void build_compression_type(::parquet::WriterProperties::Builder& builder,
                                        const TCompressionType::type& compression_type);
+
+    static std::shared_ptr<::parquet::schema::GroupNode> make_schema(const std::vector<std::string>& file_column_names,
+                                                                     const std::vector<ExprContext*>& output_expr_ctxs);
+
+    static std::shared_ptr<::parquet::WriterProperties> make_properties(const ParquetBuilderOptions& options);
+
+private:
+    static ::parquet::schema::NodePtr _make_schema_node(const std::string& name, const TypeDescriptor& type_desc,
+                                                        ::parquet::Repetition::type rep_type);
 };
 
 class FileWriterBase {
 public:
     FileWriterBase(std::unique_ptr<WritableFile> writable_file, std::shared_ptr<::parquet::WriterProperties> properties,
                    std::shared_ptr<::parquet::schema::GroupNode> schema,
                    const std::vector<ExprContext*>& output_expr_ctxs);
+
     virtual ~FileWriterBase() = default;
 
     Status init();
+
     Status write(Chunk* chunk);
+
     std::size_t file_size() const;
+
     void set_max_row_group_size(int64_t rg_size) { _max_row_group_size = rg_size; }
+
     std::shared_ptr<::parquet::FileMetaData> metadata() const { return _file_metadata; }
+
     Status split_offsets(std::vector<int64_t>& splitOffsets) const;
+
     virtual bool closed() const = 0;
 
 protected:
     virtual void _flush_row_group() = 0;
 
 private:
-    ::parquet::RowGroupWriter* _get_rg_writer();
+    class Context {
+    public:
+        Context(int16_t max_rep_level, size_t estimated_size = 0) : _max_rep_level(max_rep_level) {
+            _idx2subcol.reserve(estimated_size);
+            _def_levels.reserve(estimated_size);
+            _rep_levels.reserve(estimated_size);
```

Comment:
nit: you could init and fill the vector

Reply:
This could be a potential optimization in future.